### ISSUE TITLE
What does /upgrade do?

### ISSUE THREAD

**Issue by ioweb-gr** [MicrosoftDocs/WSL#280]
Hi, I can see in the wslconfig command there is an option /upgrade &lt;DistributionName&gt; . But there is no documentation on what it actually does apart from the Upgrades the distribution to the WslFs file system format.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: deab370f-a7a1-ed3f-89ae-22194fd8fb78
* Version Independent ID: 9f406c77-415c-80db-d596-0a7fa4ea2ba9
* Content: [Manage Linux Distributions](https://docs.microsoft.com/en-us/windows/wsl/wsl-config)
* Content Source: [WSL/wsl-config.md](https://github.com/MicrosoftDocs/WSL/blob/live/WSL/wsl-config.md)
* Service: **windows-subsystem-for-linux**
* Product: **windows-subsystem-for-linux**
* GitHub Login: @scooley
* Microsoft Alias: **scooley**

**Comment by mrdrwest** [comment]
I executed wslconfig.exe /upgrade ubuntu

It took ~10mins+ for the process to complete successfully, though I don't know exactly what it did.

**Comment by 0xabu** [comment]
Tagging @SvenGroot who probably knows

**Comment by SvenGroot** [comment]
The "/upgrade" flag converts your Linux file system from LxFs to WslFs.

As you may know, WSL has historically had two file systems: LxFs for the Linux files, and DrvFs for accessing your Windows files. As time went on, we added more features to DrvFs that moved it closer in behavior to LxFs: we added metadata support, and allowed escaping special characters that aren't legal in NTFS file names, etc. However, for various reasons (mainly different requirements), we used a different implementation for these features in DrvFs than LxFs.

WslFs is a replacement for LxFs that converges these features. It basically uses the DrvFs implementation for your internal Linux files. It's essentially DrvFs with a few options locked: metadata is always on, all directories are case sensitive, etc. It also has caching enabled, unlike DrvFs (so using your internal Linux files from Windows is still not supported, and they must be accessed through \\wsl$).

There isn't a lot of reason to upgrade to WslFs.  There may be some improved performance for some operations, but it probably won't be noticeable. In general, we only recommend WslFs for new installs, which will be turned on in a future insider build. Even more so because we're currently investigating a few strange bugs in the upgrade process, so I wouldn't recommend using this unless you're comfortable with possibly needing to reinstall the distribution if something goes wrong.

**Comment by chrants** [comment]
It would be great to have at least some tiny release notes on this. I'm like "ohh am I running something old" and so I ran upgrade without knowing what it did. You may call me an idiot. And I may be one for running options I don't know. But still, this should be documented xD

**Comment by vegerot** [comment]
I wish I had read your recommendation to not upgrade before upgrading.  @SvenGroot I've been stuck for the upgrade process for hours, and trying to run `wsl` just returns `An install, uninstall, or upgrade is in progress for this distribution.` and trying to run bash similarly gives `Cannot launch bash because an install, uninstall or upgrade is in progress for this distribution.`. 

If this is truly changing the file system I fear the restarting my computer will most certainly fubar my distro.

**Comment by vegerot** [comment]
> It would be great to have at least some tiny release notes on this. I'm like "ohh am I running something old" and so I ran upgrade without knowing what it did. You may call me an idiot. And I may be one for running options I don't know. But still, this should be documented. (￣▽￣)*

Were you able to update just fine?  For me trying to launch bash has given me the error message `Cannot launch bash because an install, uninstall or upgrade is in progress for this distribution.` for weeks

**Comment by GeekyMonkey** [comment]
Is there a way to tell which file system is currently in use?

**Comment by SvenGroot** [comment]
@GeekyMonkey Run the `mount` command in WSL, and look for the file system type of `/`, which will say either wslfs or lxfs.

Or, if you're using WSL2 (once that's released to Insiders next month), it'll say ext4.

**Comment by tjmoore** [comment]
Will there be an upgrade to ext4 or is the same advice true that it's not necessary?

**Comment by vegerot** [comment]
beta's not out yet, but I'd imagine there will be an option to upgrade to ext4.  Well actually, in WSL2 it would be either lxfs/ext4 or wslfs/ext4, since you can go back and forth between WSL1 and 2.

**Comment by ytrezq** [comment]
> TL;DR: You don't need to do this.
> 
> The "/upgrade" flag converts your Linux file system from LxFs to WslFs.
> 
> As you may know, WSL has historically had two file systems: LxFs for the Linux files, and DrvFs for accessing your Windows files. As time went on, we added more features to DrvFs that moved it closer in behavior to LxFs: we added metadata support, and allowed escaping special characters that aren't legal in NTFS file names, etc. However, for various reasons (mainly different requirements), we used a different implementation for these features in DrvFs than LxFs.
> 
> WslFs is a replacement for LxFs that converges these features. It basically uses the DrvFs implementation for your internal Linux files. It's essentially DrvFs with a few options locked: metadata is always on, all directories are case sensitive, etc. It also has caching enabled, unlike DrvFs (so using your internal Linux files from Windows is still not supported, and they must be accessed through \\wsl$).
> 
> There isn't a lot of reason to upgrade to WslFs. There may be some improved performance for some operations, but it probably won't be noticeable. In general, we only recommend WslFs for new installs, which will be turned on in a future insider build. Even more so because we're currently investigating a few strange bugs in the upgrade process, so I wouldn't recommend using this unless you're comfortable with possibly needing to reinstall the distribution if something goes wrong.

@SvenGroot I accidentaly deregistered my Legacy installation at `%APPDATA%\local\lxss` through deleting the resulting registry keys leaving the files no longer recognized. How to upgrade to WslFs?
